### PR TITLE
[IOTDB-1948] [IOTDB-1949] Group by fill supports vector

### PR DESCRIPTION
**Pull request by CRZbulabula**

### CODE REVIEW

**Review by ericpai — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByFillDataSet.java [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+    } catch (QueryProcessException e) {
+      // ignored
+    }
+
+    return record;
+  }
+
+  private void fillRecord(int index, RowRecord record) throws IOException {
+    // Don't fill count aggregation
+    if (Objects.equals(aggregations.get(index), "count")) {
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByFillDataSet.java:189]: When some time range contains nothing, the "count" aggregation result is not null, but 0 in IoTDB. 0 means 0 in "count" aggregation, so Brother Dong and Mr.Qiao decided that we shouldn't fill "count" result.

File: integration/src/test/java/org/apache/iotdb/db/integration/IoTDBGroupByFillMixPathsIT.java [hunk 2]
```diff
@@ -122,367 +127,361 @@ private void prepareData() {
...
+          "37,126.0,27,36",
+          "42,126.0,27,36",
+          "47,135.0,22,51",
+          "52,110.0,20,56",
+          "57,110.0,20,56",
+          "62,110.0,20,56"
+          // "72,33.0,33,72"
+        };
+
+    try (Connection connection =
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply integration/src/test/java/org/apache/iotdb/db/integration/IoTDBGroupByFillMixPathsIT.java]: done

File: integration/src/test/java/org/apache/iotdb/db/integration/IoTDBGroupByFillMixPathsIT.java [hunk 3]
```diff
@@ -122,367 +127,361 @@ private void prepareData() {
...
+        assertEquals(retArray.length, cnt);
+      }
+    } catch (Exception e) {
+      e.printStackTrace();
+      fail(e.getMessage());
+    }
+  }
+
+  @Test
+  public void MultiPathsMixTest() {
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply integration/src/test/java/org/apache/iotdb/db/integration/IoTDBGroupByFillMixPathsIT.java]: done

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByFillDataSet.java [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+      if (nextTVLists.get(i).getTime(nextIds[i]) == curTimestamp) {
+        previousTimes[i] = curTimestamp;
+        previousValues[i] = getNextCacheValue(i);
+        nextIds[i] += 1;
+        nextTVLists.get(i).setEvictionUpperBound(nextIds[i]);
+        slideSet.set(i);
+      }
+    }
+
+    while (slideSet.cardinality() > 0 && dataSet.hasNextWithoutConstraint()) {
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByFillDataSet.java:319]: Yes, it's my mistake, and I fixed it now

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByFillDataSet.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ericpai) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByFillDataSet.java:186]: I think it's better to leave the choice to users, as we don't know what's the real meaning of `count()` in their business., although in IoTDB `count(s) == null` is equal to  `count(s) == 0`. But if we strictly fill a `0` to the `count `result, then there's no other opportunities to let user fill with other options, and the users may be confused by the results (they got `null` without fill but got `0` with any fill types).

Another solution is to let `count` aggregation never return `null`. It's better than this work around method. We only need to make `CountAggregationResult.hasCandidateResult()` return `true`.
Here's another interesting question. If any time window is empty, which aggregations should return a real value,   not `null`? Currently I think `sum` and `count` should return 0.

How do you think about this @jixuan1989 ?

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByFillDataSet.java [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+  public RowRecord nextWithoutConstraint() throws IOException {
+    if (!hasCachedTimeInterval) {
+      throw new IOException(
+          "need to call hasNext() before calling next() " + "in GroupByFillDataSet.");
+    }
+
+    hasCachedTimeInterval = false;
+    RowRecord record;
+    long curTimestamp;
+    if (leftCRightO) {
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByFillDataSet.java:153]: see IoTDBGroupByFillIT.leftORightCPreviousTest

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByFillDataSet.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByFillDataSet.java:186]: Currently, users never get `null` in `count` aggregation. As I said, the `count` result is 0 when the time range contains nothing. I will discuss that question with other contributers in these days!

[the remark this answers is not in the document]
Comment (ericpai) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByFillDataSet.java:186]: If it's never null, the `fillRecord ` is never called as well, so is this special code  useless?

[the remark this answers is not in the document]
Comment (CRZbulabula) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByFillDataSet.java:186]: Yes, it is. I'll remove it tomorrow. And for `sum` aggregation, `0` and `null` have different meanings. The former can be a sum of 0 in the time range, while the latter means that there is no data in the time range. So currently, we distinguish `0` and `null` in `sum` aggregation, and we fill `null` in `sum`. You can see this on [IoTDB-1884](https://github.com/apache/iotdb/pull/4233).

[the remark this answers is not in the document]
Comment (ericpai) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByFillDataSet.java:186]: Make sense~

**Review by ericpai — APPROVED** [review]

**Review by JackieTien97 — APPROVED** [review]

**Pull request merged**